erlcloud_aws changes, merging code from alertlogic repo

alexturkin opened the pull request:
this PR is part and preparation for https://github.com/gleber/erlcloud/pull/215

the purpose of this PR is mostly to commit common changes from `erlcloud_aws` module which will simplify further PRs to come

### What's changed? 
* adding x-amz-content-sha256 to header for signv4 (it will be needed for s3 related changes in further s3-related PRs)
* refactoring: move common code for retrieving region for host to one place
* https://github.com/alertlogic/erlcloud/pull/20 - added Reporting of AWS statuses using

All further PRs are expected to be as one PR per each AWS service.

ransomr commented on src/erlcloud_aws.erl:
```diff
@@ -194,15 +202,23 @@ format_timestamp({{Yr, Mo, Da}, {H, M, S}}) ->
-            #aws_config{access_key_id=os:getenv("AWS_ACCESS_KEY_ID"),
-                        secret_access_key=os:getenv("AWS_SECRET_ACCESS_KEY")};
+            AccessKeyId = case os:getenv("AWS_ACCESS_KEY_ID") of
+                              false -> "";
```
Is there a reason that access_key_id needs to be a string? It seems like `undefined` would be a more conventional value if it isn't set, and would be less likely to break things like guard tests that look for lists.

alexturkin replied:
agree. I'm not sure there was specific reason for that. Anyway, tests are fine with that both in your repo and ours, so I've changed this to be undefined

ransomr commented on src/erlcloud_aws.erl:
```diff
@@ -409,9 +424,54 @@ signing_key(Config, Date, Region, Service) ->
+                        _ -> false
+                    end
+                end, 
+            ServiceNames)
```
please fix indent - `ServiceNames` and `Statuses` should match corresponding `fun`

alexturkin replied:
ok

alexturkin commented:
@ransomr I've addressed your comments, please take a look.

We have a bunch of branches coming up with a lot of other changes, but they're based on this ones, so after this will be merged, there will be a lot of other PRs :)

ransomr commented:
Thank you very much for the contributions - looking forward to the rest of it.
